File: Medium_Python/5_Longest_Palindromic_substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        maxlen=1
        if len(s)==0:
            return ""
        tempstring=s[0]
        for i in range(len(s)):
            left,right=i-1,i+1
            while left>=0 and right<len(s):
                if s[left]!=s[right]:
                    break
                else:
                    left-=1
                    right+=1
            
            if right-left-1>maxlen:
                maxlen=right-left-1
                tempstring=s[left+1:right]
            left,right=i-1,i+2
            if i+1<len(s) and s[i]==s[i+1]:
                while left>=0 and right<len(s):
                    if s[left]!=s[right]:
                        break
                    else:
                        left-=1
                        right+=1
                if right-left-1>maxlen:
                    maxlen=right-left-1
                    tempstring=s[left+1:right]
        return tempstring

File: Medium/5_longest_palindromic_substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
    //DP
        vector<vector<bool> >dp(s.length(),vector<bool>(s.length(),true));
        int maxlen = 1;
        int pos = 0;
        for(int j = 1; j < s.length();j++)
        {
            for(int i = 0; i < j;i++)
            {
                if(s[i]!=s[j])
                {
                    dp[i][j]=false;
                }
                else
                {
                        if(dp[i+1][j-1]==true)
                        {
                            
                            dp[i][j]=true;
                            if(j-i+1>maxlen)
                        {
                            maxlen = j-i+1;
                            pos = i;
                        }
                        
                    }
                    else
                    {
                        dp[i][j]=false;
                    }
                    
                }
                
            }
        }

        return s.substr(pos,maxlen);
    }
};
